crates/services/sync/src/import.rs

```diff
@@ -0,0 +1,363 @@
+    state: SharedMutex<State>,
+    /// Notify import when sync has new work.
+    notify: Arc<Notify>,
+    /// Configuration parameters.
+    params: Config,
+    /// Network port.
+    p2p: Arc<P>,
+    /// Executor port.
+    executor: Arc<E>,
+    /// Consensus port.
+    consensus: Arc<C>,
+}
+
+impl<P, E, C> Import<P, E, C> {
+    pub(crate) fn new(
+        state: SharedMutex<State>,
+        notify: Arc<Notify>,
+        params: Config,
+        p2p: Arc<P>,
+        executor: Arc<E>,
+        consensus: Arc<C>,
+    ) -> Self {
+        Self {
+            state,
+            notify,
+            params,
+            p2p,
+            executor,
+            consensus,
+        }
+    }
+}
+impl<P, E, C> Import<P, E, C>
+where
+    P: PeerToPeerPort + Send + Sync + 'static,
+    E: BlockImporterPort + Send + Sync + 'static,
+    C: ConsensusPort + Send + Sync + 'static,
+{
+    pub(crate) async fn import(
+        &self,
+        shutdown: &mut StateWatcher,
+    ) -> anyhow::Result<bool> {
+        self.import_inner(shutdown).await?;
+
+        Ok(wait_for_notify_or_shutdown(&self.notify, shutdown).await)
+    }
+
+    async fn import_inner(&self, shutdown: &StateWatcher) -> anyhow::Result<()> {
+        // If there is a range to process, launch the stream.
+        if let Some(range) = self.state.apply(|s| s.process_range()) {
+            // Launch the stream to import the range.
+            let (count, result) = self.launch_stream(range.clone(), shutdown).await;
+
+            // Get the size of the range.
+            let range_len = range.size_hint().0 as u32;
+
+            // If we did not process the entire range, mark the failed heights as failed.
+            if (count as u32) < range_len {
+                let range = (*range.start() + count as u32)..=*range.end();
+                self.state.apply(|s| s.failed_to_process(range));
+            }
+            result?;
+        }
+        Ok(())
+    }
+
+    /// Launches a stream to import and execute a range of blocks.
+    ///
+    /// This stream will process all blocks up to the given range or
+    /// an error occurs.
+    /// If an error occurs, the preceding blocks still be processed
+    /// and the error will be returned.
+    async fn launch_stream(
+        &self,
+        range: RangeInclusive<u32>,
+        shutdown: &StateWatcher,
+    ) -> (usize, anyhow::Result<()>) {
+        let Self {
+            state,
+            params,
+            p2p,
+            executor,
+            consensus,
+            ..
+        } = &self;
+        // Request up to `max_get_header_requests` headers from the network.
+        get_header_range_buffered(range.clone(), params, p2p.clone())
+        .map({
+            let p2p = p2p.clone();
+            let consensus_port = consensus.clone();
+            move |result| {
+                let p2p = p2p.clone();
+                let consensus_port = consensus_port.clone();
+                async move {
```

Comment:
Hmm, it seems we can create this function one time during the creation of the task and use it here. Instead of creating it on each iteration.

But it will affect your current code, so I guess it is enough for now to create it at the begin of the function.

Reply:
Is that really a concern though? Very likely this just gets optimized away by the compiler.

Reply:
Also we can't do that because this takes the closure owned so you would need to clone the closure anyway.